### PR TITLE
feat: windows msvc toolchain support

### LINKED ISSUE #443: [CI] Run some integration tests on Windows
Now that we're close to an MVP, we should start thinking about making sure `lux` works on Windows.

### PR DESCRIPTION
**Pull request by mrcjkb**
Stacked on #512, #584.

Closes #443.

### CODE REVIEW

**Comment by mrcjkb** [comment]
Current status: luafilesystem is failing to compile with

```
error LNK2019: unresolved external symbol __imp_timeBeginPeriod referenced in function time_lua_sleep\r\n6d5d9cb306644220-time.o : error LNK2019: unresolved external symbol __imp_timeEndPeriod referenced in function time_lua_sleep\r\n6d5d9cb306644220-random.o : error LNK2019: unresolved external symbol BCryptGenRandom referenced in function lua_get_random_bytes\r\nC:\\Users\\RUNNER~1\\AppData\\Local\\Temp\
```

This indicates that the

```lua
win32 = { "advapi32", "winmm", "bcrypt" },
```

platform override either isn't being picked up or the libraries are being passed to the wrong `Build` command.

**Comment by mrcjkb** [comment]
luasystem and luafilesystem seem to build successfully now, but lua-term fails, I think because it's using posix functions that exist as aliases in msvc, but are deprecated.
I'm disabling the msvc busted tests for now - we can tackle that another time.

**Comment by mrcjkb** [comment]
Looks like we're missing a variable substitution.

```
-- LUA_LIBFILE: $(LUALIB)
```

File: lux-lib/src/build/utils.rs [hunk 1]
```diff
@@ -113,56 +107,122 @@ pub(crate) fn compile_c_files(
...
         .cargo_metadata(false)
-        .cargo_debug(false)
         .cargo_warnings(false)
-        .debug(false)
+        .warnings(false)
         .files(files)
         .host(std::env::consts::OS)
-        .opt_level(3)
+        .include(&lua.include_dir)
+        .opt_level(2)
```
Comment (vhyrro) [inline lux-lib/src/build/utils.rs:115]: question: what's the rationale for lowering the opt level? Does it cause problems on windows, was this temporarily used for easier debugging of compiler output or did it cause clashes with our default cflags (which use O2)?

File: lux-lib/src/build/utils.rs [hunk 2]
```diff
@@ -113,56 +107,122 @@ pub(crate) fn compile_c_files(
...
         .out_dir(intermediate_dir)
         .target(&host.to_string());
 
-    for arg in lua.compile_args() {
+    let compiler = build.try_get_compiler()?;
+    let is_msvc = compiler.is_like_msvc();
```
Comment (vhyrro) [inline lux-lib/src/build/utils.rs:120]: nitpick: useless variable (isn't used lower down on line 137)

File: lux-lib/src/build/utils.rs [hunk 3]
```diff
@@ -237,39 +299,70 @@ pub(crate) fn compile_c_modules(
...
-        .args(library_args)
-        .output()?;
+    let output_path = parent.join(&file);
+    let output = if is_msvc {
+        let def_temp_dir = tempdir::TempDir::new("msvc-def")?.into_path().to_path_buf();
+        let def_file = mk_def_file(def_temp_dir, &file, target_module)?;
+        build
+            .try_get_compiler()?
+            .to_command()
+            .arg("/nologo")
```
Comment (vhyrro) [inline lux-lib/src/build/utils.rs:335]: question: I see refactors from `/nologo` to `/NOLOGO`, should this also be changed?

File: lux-lib/src/build/variables.rs [hunk 4]
```diff
@@ -130,4 +130,12 @@ mod tests {
...
         assert_eq!(result, expected);
     }
+
+    #[test]
+    fn empty_string() {
+        assert_eq!(
+            substitute(&[&TestVariables], "$(EMPTY_STRING)").unwrap(),
+            "".to_string()
+        );
+    }
```
Comment (vhyrro) [inline lux-lib/src/build/variables.rs:140]: todo: we should also create a test for a truly empty string: `substitute(&[&TestVariables], "")`, as that was also part of the previous related bug report.

File: lux-workspace-hack/Cargo.toml [hunk 5]
```diff
@@ -46,8 +46,6 @@ smallvec = { version = "1", default-features = false, features = ["const_new"] }
 time = { version = "0.3", default-features = false, features = ["formatting", "macros", "parsing"] }
 tokio = { version = "1", features = ["full"] }
 tokio-util = { version = "0.7", features = ["codec", "io"] }
-toml = { version = "0.8" }
-toml_edit = { version = "0.22", features = ["serde"] }
```
Comment (vhyrro) [inline lux-workspace-hack/Cargo.toml:50]: question: is this removal supposed to happen?

**Review by vhyrro — CHANGES_REQUESTED** [review]
:tada: amazing work, I can tell this was really painful just by reading the code. Just some questions, nitpicks and a TODO.

File: lux-lib/src/build/utils.rs (hunk 1 above)
Comment (mrcjkb) [reply lux-lib/src/build/utils.rs]: That was left over from troubleshooting.

File: lux-workspace-hack/Cargo.toml (hunk 5 above)
Comment (mrcjkb) [reply lux-workspace-hack/Cargo.toml:50]: This was generated by `cargo hakari generate`.

**Comment by mrcjkb** [comment]
Windows test failure is the flaky tree-sitter parser build.

**Review by vhyrro — APPROVED** [review]
